Use custom gRPC dialer to override default proxy dialer

dani-docker opened the pull request:
Signed-off-by: Dani Louca <dani.louca@docker.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/swarmkit/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Followed the same logic as in https://github.com/docker/swarmkit/pull/2419 to use a custom grpc dialer and prevent the use of default dialer which picks up the proxy env variable when a manager is joining the cluster.
The affected method `dial` is only used when manager is joining the cluster  and part of `checkHealth` which is also called on join.

**- How to test it**
- Follow the link to configure 2 nodes (not part of swarm cluster) to use HTTP Proxy https://docs.docker.com/config/daemon/systemd/#httphttps-proxy
- Enable swarm on `node 1` , `docker swarm init`
- Join `node 2` as manager

Actual Result:
```
[centos@esc-1003-1 ~]$ docker swarm join --token SWMTKN-1-2v3cay35o2yavb2bqge6wvyhg6esdn47g82ydi534npgcd4cym-4rmvy3uo2acmbenjvmf8tkx52 10.10.13.91:2377
Error response from daemon: manager stopped: can't initialize raft node: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp: lookup proxy.example.com on 10.10.0.2:53: no such host"
```

Expected Result (with this PR):
Node joins successfully 
```
[root@esc-1003-1 ~]# docker swarm join --token SWMTKN-1-2v3cay35o2yavb2bqge6wvyhg6esdn47g82ydi534npgcd4cym-4rmvy3uo2acmbenjvmf8tkx52 10.10.13.91:2377
This node joined a swarm as a manager.
```

**- Description for the changelog**
Fix manager node join when HTTP_PROXY, HTTPS_PROXY, or NO_PROXY are used

dperny commented:
LGTM. Pretty sure the test failure is known so I'm rebuilding.

thaJeztah commented:
Should we have an integration test for this that set an [invalid domain](https://tools.ietf.org/html/rfc2606) (`HTTPS_PROXY=myproxy.invalid`) as a proxy, then tries to join?

thaJeztah commented:
Looks like this is related to https://github.com/moby/moby/issues/36951 and https://github.com/moby/moby/issues/35395#issuecomment-341713523

trapier commented:
We still waiting on CI here?  Current failure doesn't look related.

thaJeztah commented:
Not a maintainer here, but ideally;

- May be good to update the commit message, and include a bit more context about the reason for this change
- Add an integration test (if possible) to verify the behavior https://github.com/docker/swarmkit/pull/2802#issuecomment-455662961

I prepared backports, but those will have to be updated if this PR is updated before it's merged


I also triggered CI, but it's now failing on

```
gometalinter: error: unknown linters: gosimple
make: *** [check] Error 1
Exited with code 2
```

which should be fixed by https://github.com/docker/swarmkit/pull/2811 after rebasing this PR

wk8 commented:
Agreed an integration test would be great :)

anshulpundir commented:
+1 for https://github.com/docker/swarmkit/pull/2802#issuecomment-462290709

thaJeztah commented:
ping @dani-docker can you address the review comments?

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "esc-1003" git@github.com:dani-docker/swarmkit.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842357316928
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

dani-docker commented:
@thaJeztah 
I rebased and updated the comment as requested.
As for the test case, this issue completes the fix in https://github.com/docker/swarmkit/pull/2419 which is also missing a test, I will look into a test shortly
